### PR TITLE
Move websocket header normalization later

### PR DESCRIPTION
**Pull request by nfreya**
A regression since #6621, these headers can be rewritten by `cloneRequest` and by `req.Headers.Add` from e.g. user and transport ops. Moving this closer to where it's round-tripped feels like the nicest solution.

I've put the harness I used to reproduce the issue at
  https://codeberg.org/nfreya/caddy-websocket-headers-repro

I encountered this issue while trying to reverse proxy the iKVM from an old SuperMicro BMC which was using some ancient noVNC fork, and I had a suspicion that the 400 response from the upstream was due to `CanonicalMIME`-related normalization combined with the sort of naive `strstr` you'd expect in the C code of a firmware written under ... 'constraints.' From there I found the comment referencing #6621 and the open #7784. Since that more recent one has AI usage, I didn't look further into it on principle, and from skimming the corresponding #7786 now, it looks uglier than my approach of just moving the normalize down. Any unintended consequences of this movement would be limited to consumers of the headers that branch on case difference for a request between the old and new `normalizeWebsocketHeaders` call site, and that seems like a rare enough edge case with obviously unsupported semantics to be worth breaking for simplicity's sake. If some unit tests of this behaviour are necessary for acceptance we can do that, but the regression itself was caused in a way that seems difficult to guard against by anything other than relatively fragile integration testing, so I've left it at this.

`git log -S` showed me #7911 affecting something tangential since those allocations come from the same mismatch between `CanonicalMIME` and the original websocket RFC -- this PR doesn't make that situation strictly worse, as leaving these headers in `CanonicalMIME` form before the request is cloned however many times `reverse_proxy` ends up cloning it makes it less likely for that allocation to occur, but I also *think* the `normalizeWebsocketHeaders` method is unlikely do any allocations since it's just migrating a (pointer-to-const key, slice value) from one hashmap slot to another (unverified). The extra `upgradeType` call is in the path of every request, but memoizing that seems annoying just to eliminate what is just a scan of the map when there's so much other code in this path.

## Assistance Disclosure

No AI was used.

### CODE REVIEW

**Comment by francislavoie** [comment]
Seems to me like https://github.com/caddyserver/caddy/pull/7786 should cover this already

**Comment by steadytao** [comment]
I think this placement is cleaner than 7786: it normalises once immediately before reverseProxy, after the request-header changes and only for upgrade requests. Please bring over a regression test which exercises this actual proxy-loop path and fails when the call is removed; with that coverage I would prefer this PR.

**Comment by nfreya** [comment]
Took a bit to get back to this, sorry. There's now an integration test for the WebSocket header normalization which fails if the first commit of the branch is removed. This catches the #7784 `header_up` case explicitly, which is not exactly the config I had with the https upstream, but they're equivalent as they both modify Headers after the unpatched `normalizeWebsocketHeaders` site, and `header_up` makes for an easier test to set up since it doesn't need TLS serving.

File: caddytest/integration/reverseproxy_test.go [hunk 1]
```diff
@@ -832,6 +835,84 @@ func TestReverseProxySNIPlaceHolder(t *testing.T) {
...
+		t.Fatalf("failed to listen: %v", err)
+	}
+	// http.Server would canonicalize these headers, so plain tcp:
+	defer ln.Close()
+	go func() {
+		for {
+			conn, err := ln.Accept()
+			if err != nil {
+				if !errors.Is(err, net.ErrClosed) {
+					t.Fatalf("failed to accept: %v", err)
```
Comment (steadytao) [inline caddytest/integration/reverseproxy_test.go]: Do not call `t.Fatalf` from this goroutine and do not assume one TCP read contains the complete request headers. Accumulate into a bounded buffer until `\\r\\n\\r\\n`, send errors through a channel and assert them from the test goroutine.

**Review by steadytao — CHANGES_REQUESTED** [review]
No need to be sorry. Thank you very much for your work. Just a issue in goroutine handling but otherwise very lovely!

File: caddytest/integration/reverseproxy_test.go [hunk 2]
```diff
@@ -832,6 +835,105 @@ func TestReverseProxySNIPlaceHolder(t *testing.T) {
...
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("failed to listen: %v", err)
+	}
+	// http.Server would canonicalize these headers, so plain tcp:
+	defer ln.Close()
+	errc := make(chan error, 1)
+	go func() {
+		for {
+			conn, err := ln.Accept()
```
Comment (steadytao) [inline caddytest/integration/reverseproxy_test.go:877]: Could each accepted connection be handled in a small helper with `defer conn.Close()` ?

Every read, size-limit or write error currently returns from the server goroutine without closing the accepted socket; only the successful path reaches the explicit Close. Please also avoid deferring directly inside this outer accept loop since that would retain successful connections until the listener exits.

**Review by steadytao — CHANGES_REQUESTED** [review]
Looking really good!

File: caddytest/integration/reverseproxy_test.go (hunk 2 above)
Comment (nfreya) [reply caddytest/integration/reverseproxy_test.go:877]: Always Closing conn now, but no longer checking the error from Close as getting it out of the defer feels extra for a test. I also moved from plain bytes to bufio and textproto since it declares the intent better. There's also now a defer to empty errc with a close on the server side to prevent that blocking the server goroutine. An external client can still break the test, this is just to make it obvious if some odd change to httpproxy decides to hit the upstream in an unexpected way.

**Review by steadytao — APPROVED** [review]
Thank you very much, freya. LGTM.

**Comment by steadytao** [comment]
Pending #7964 -- I hate CI :D

**Review by mholt — APPROVED** [review]
